language/diem-vm/parallel-executor/src/task.rs

```diff
@@ -0,0 +1,78 @@
+
+use anyhow::Result;
+use mvhashmap::{MVHashMapView, Version};
+use std::{fmt::Debug, hash::Hash};
+
+/// The execution result of a transaction
+#[derive(Debug)]
+pub enum ExecutionStatus<T, E> {
+    /// Transaction was executed successfully.
+    Success(T),
+    /// Transaction hit a none recoverable error during execution, halt the execution and propagate
+    /// the error back to the caller.
+    Abort(E),
+    /// Transaction was executed successfully, but will skip the execution of the trailing
+    /// transactions in the list
+    SkipRest(T),
+    /// Transaction has an unexpected read dependency that is blocked by `Version` transaction.
+    /// Put the transaction back to the scheduler.
+    Retry(/*blocked_by*/ Version),
+}
+
+/// Trait that defines a transaction that could be parallel executed by the scheduler. Each
+/// transaction will write to a key value storage as their side effect.
+pub trait Transaction: Clone + Sync + Send + 'static {
+    type Key: PartialOrd + Send + Sync + Clone + Hash + Eq;
+    type Value: Send + Sync;
+}
+
+/// Trait for inferencing the read and write set of a transaction.
+pub trait TransactionInferencer: Sync {
+    /// Type of transaction and its associated key.
+    type T;
+    type Key;
+
+    /// Get the read set of a transaction. Read set estimation is used simply to improve the
+    /// performance by exposing the read dependencies. Imprecise estimation won't cause execution
+    /// failure.
+    fn infer_reads(&self, txn: &Self::T) -> Result<Vec<Self::Key>>;
+
+    /// Get the write set of a transaction. Write set estimation is crucial to the execution
+    /// correctness as there's no way to resolve read-after-write conflict where a write is
+    /// unexpected. Thus we require write to be an over approximation for now.
+    fn infer_writes(&self, txn: &Self::T) -> Result<Vec<Self::Key>>;
+}
+
+/// Trait for single threaded transaction executor.
+// TODO: Sync should not be required. Sync is only introduced because this trait occurs as a phantom type of executor struct.
+pub trait ExecutorTask: Sync {
+    /// Type of transaction and its associated key and value.
+    type T;
+    type Key: PartialOrd + Send + Sync + Clone + Hash + Eq;
+    type Value: Send + Sync;
+
+    /// The output of a transaction. This should contain the side effect of this transaction.
+    type Output: Default + Send + Sync;
+
+    /// Type of error when the executor failed to process a transaction and needs to abort.
+    type Error: Clone + Send + Sync;
+
+    /// Type to intialize the single thread transaction executor. Copy and Sync are required because
+    /// we will create an instance of executor on each individual thread.
+    type Argument: Sync + Copy;
+
+    /// Create an instance of the transaction executor.
+    fn init(args: Self::Argument) -> Self;
+
+    /// Execute one single transaction given the view of the current state.
+    fn execute_transaction(
+        &self,
```

Comment:
Either should be fine actually. This structure will be spawned on each individual thread so it doesn't matter if it's `mut` or not. `&mut` will actually be a bit more flexible but `&self` seems like the right api to me because we don't want the user to rely on the fact that it is uniquely shared by each individual thread.